Lock transitive Node dependencies (#413)
BSI-Business-Systems-Integration-AG/org.eclipse.scout.docs: docs/modules/technical-guide/pages/user-interface/build-stack.adoc

[#node-dependency-resolution]
=== Dependency Resolution

https://pnpm.io/[Pnpm] takes care to link all dependencies declared in a `package.json` file to a `node_modules` directory next to it. If it is a 3rd party dependency, it is downloaded to a local store before.

The dependency version to use is part of the dependency declaration in the `package.json` and follows the https://semver.org/[SemVer] specification.
This specification allows to declare range dependencies which might match several versions (even versions in the future which have not yet been published).

> When a dependency is installed, pnpm decides which version to use based on the SemVer constraint declared.

Reviewer: based on the declared SemVer constraint.

Using ranged versions has the benefit that you automatically get newer versions e.g. including bugfixes without changing anything in your dependency declarations.
On the other hand you also get regressions automatically, and it may prevent reproducible build outputs.

> Furthermore, there have been several malicious dependency updates of packages whose developer accounts have been hacked.

Reviewer: Maybe rewrite the last sentence include the wording "supply chain attack". 
E.g: It also increases the risk of supply chain attacks, where an attacker compromises an npm account and publishes a malicious package that will be downloaded automatically.

> To address these issues Scout 26.2 or newer by default locks most of the ranged dependencies to a fixed version.

Reviewer: ```suggestion
To address these issues, Scout 26.2 or newer locks most of the ranged dependencies to a fixed version by default.
```

Excluded are `snapshot` dependencies which should always point to the newest available snapshot version.

> Because Scout only needs partial locking, the pnpm default locking mechanism using a `pnpm-lock.yaml` file cannot be used (because this always locks the full dependency tree).

Reviewer: ```suggestion
Excluded are `snapshot` dependencies which should always point to the newest available snapshot version.
These `snapshot` dependencies are the reason Scout cannot use pnpm's default locking mechanism based on a `pnpm-lock.yaml` file. The lock file always pins the entire dependency tree, whereas snapshot dependencies must not be locked.
```

Instead, the locked versions are written to the https://pnpm.io/settings#overrides[pnpm overrides] in the `pnpm-workspace.yaml` file.
The new `install` command of the `@eclipse-scout/releng` module generates these overrides when dependencies are installed locally during development.

> The Maven build scripts only use the available overrides and therefore exactly use the versions specified.

Reviewer: ```suggestion
The Maven build scripts only use the available overrides and therefore exactly use the specified versions.
```

The `install` command of the `@eclipse-scout/releng` module has the following parameters:

* `updateMode`: Specifies which dependencies should be updated during install. Default is `snapshots`. Must be one of the following:

> ** `required`: Update dependencies that do not fulfill the version declaration in the `package.json`. This should be used during development so that the overrides are adapted as required in case you modify the dependencies in the `package.json`.

Reviewer: Maybe add some examples for each mode to make it more clear what happens

> ** `required`: Update dependencies that do not fulfill the version declaration in the `package.json`. This should be used during development so that the overrides are adapted as required in case you modify the dependencies in the `package.json`.

Reviewer: I was wondering, why certain dependencies were not included in scout-overrides. I changed the jQuery version and nothing happened even if I used updateMode required. Then I noticed jQuery is not included in the yaml. Maybe describe how the scout-overrides are created?

** `possible`: Update all dependencies to the newest possible version that currently exists. This can be used if you want to perform a manual full update of everything.

> ** `snapshots`: Only `snapshot` dependencies are updated. All other dependencies use the version declared in the overrides. As only `snapshots` can be updated, and they are never part of the overrides, the overrides will not be touched in this mode.

Reviewer: Maybe add that snapshots mode is typically used by CI builds

* `convergenceLogLevel`: Specifies which non-convergent dependencies should be logged. Default is `own`. Must be one of the following:
** `all`: Log all dependencies for which several versions are in use (non-unique versions).

> ** `single-external`: Log dependencies for which a single pnpm-workspace-external dependency exists. This may be handy to eliminate duplicate dependencies by adapting the workspace to the single external one.

Reviewer: I don't really understand the single-external log-level, maybe you can provide an example? 
Also, it would be useful to know when I need to use these log levels and how, maybe add an extra chapter

** `own`: Log only mixed package versions in the own pnpm-workspace. These may be misconfigured (old) packages.
** `none`: Nothing is logged.

If you want to disable dependency locking using overrides and restore the old (and riskier) behavior do the following:

. Remove the overrides from the `pnpm-workspace.yaml` if existing.
. Replace the `pnpx @eclipse-scout/releng install` command in the `pnpm-install` script of your `package.json` files with the following: `pnpm install --ignore-scripts`.
. This will generate the `pnpm-lock.yaml` file again when executed.

> .. If you want to lock the whole tree: commit it to your VCS.

Reviewer: ```suggestion
.. If you want to lock the whole tree (including snapshot): commit it to your VCS.
```
